action_server/src/action_server/test_tools.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import print_function
+import collections
+from nose.tools import nottest
+
+from grammar_parser.cfgparser import CFGParser, Conjunct, Option, Rule
+from robot_skills.mockbot import Mockbot
+from action_server.actions.action import ConfigurationResult
+from action_server.task_manager import TaskManager
+
+TMTestResult = collections.namedtuple("TMTestResult", ["recipe", "config_result"])
+
+
+@nottest
+def test_grammar(grammar, grammar_target):
+    """
+    Tests grammars against the capabilities of the action server, to be used in unit tests of challenges.
+
+    Based on the actions rule, it recursively determines a viable option (e.g., 'go to the couch_table'), parses
+    this using the CFGParser and tries to configure the task manager based on the parse result. All results (even
+    if the configuration step with the task manager raises an exception) are stored in a result dict. Eventually,
+    all positive results are printed and if negative results have occurred, an AssertionError is raised with a
+    message containing all failed options.
+
+    :param grammar: grammar that is used
+    :param grammar_target: top level grammar_target of the grammar
+    :raises: AssertionError if the test has failed
+    """
+    # Create variables
+    parser = CFGParser.fromstring(grammar)
+    task_manager = TaskManager(robot=Mockbot())
+    actions_rule = _find_actions_rule(parser, grammar_target)
+
+    test_results = {}
+    assert actions_rule.options, \
+        "There are no options in the actions rule. Therefore, there doesn't seem anything to test: that can't be right"
+    for option in actions_rule.options:  # type: Option
+
+        result_str = ""
+
+        # For each conjunct, we want to have a viable option
+        for conjunct in option.conjuncts:  # type: Conjunct
+            result_str += _resolve_conjunct(parser, conjunct.name)
+        result_str = result_str.rstrip(" ")  # type: str
+
+        # Now, we can parse the result string to get the action description
+        actions_definition = parser.parse_raw(
+            target=grammar_target,
+            words=result_str,
+        )  # type: dict
+        assert "actions" in actions_definition, "'actions' not in actions definition, don't know what to do"
+
+        # Test the action description with the task manager
+        try:
+            test_result = task_manager.set_up_state_machine(
+                recipe=actions_definition["actions"],
+            )  # type: ConfigurationResult
+        except Exception as e:
+            test_result = ConfigurationResult()
+            test_result.message = "Configuration crashed: {}".format(e.message)
```

Comment:
Add full traceback. As current messages are only last line and don't give a clue about where it actually happened.